nanofi/examples/tail_file.c

```diff
@@ -0,0 +1,270 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+    *
+    *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+#include "api/nanofi.h"
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <limits.h>
+#include <sys/stat.h>
+
+typedef struct flow_file_records {
+    flow_file_record ** records;
+    uint64_t len;
+} flow_file_records;
+
+typedef struct file_offset {
+    int offset;
+} file_offset;
+
+struct flow_file_records * flowfiles = NULL;
+struct file_offset * fileoffset = NULL;
+struct nifi_instance * instance = NULL;
+
+void transmit_flow_files(nifi_instance * instance) {
+    if (flowfiles) {
+        for (int i = 0; i < flowfiles->len; ++i) {
+            if (flowfiles->records[i]) {
+                transmit_flowfile(flowfiles->records[i], instance);
+            }
+        }
+    }
+}
+
+char * get_string(char * begin, char * end) {
+
+    if (begin == NULL || (end != NULL && end < begin)) {
+        return NULL;
+    }
+
+    int len = 0;
+    char * buff = NULL;
+
+    if (begin == end) {
+        buff = (char *)malloc(2 * sizeof(char));
+        strncpy(buff, begin, 1);
+        buff[1] = '\0';
+        return buff;
+    }
+
+    if (end) {
+        len = end - begin;
+        buff = (char *)malloc((len+1) * sizeof(char));
+        strncpy(buff, begin, len);
+        buff[len] = '\0';
+    }
+    else {
+        len = strlen(begin);
+        buff = (char *)malloc((len+1) * sizeof(char));
+        strncpy(buff, begin, len);
+        buff[len] = '\0';
+    }
+    return buff;
+}
+
+void split_string(char * str, char delim, char ** collector, int * num_strings)
```

Comment:
This is also a cool a powerful utility for nanofi, so I would prefer to have it in core.
Please make the source const.

To return multiple values the current way we do it in Nanofi is to return a struct that contains these (for eg. a ptr and the corresponding size). Please note that this is only my preference, feel free to convince me about the opposite way. 
In case you prefer to stay with the current signature please make sure to add comments to differentiate in and out args.

Reply:
I would also like to this either documented or changed.
I have a few issues with this function:
 - the contract is not clear (you should preallocate collector, and preallocate it so that it could accommodate all potential substrings, and num_strings have to be zeroed before passing it to the function)
 - preallocating the collector so that it could store all potential substrings is wasteful. As we have the whole string in memory I would prefer for this function to iterate through the string, count the substrings, and allocate the return array itself.
 - based on its usage, duplicating the substrings is not needed. Returning with an array of offsets and lengths would be more efficient.
 - if we want to keep the source string unmodified, the current implementation is fine, but if we are ok with modifing it, the whole implementation could be replaced with strtok

Reply:
@bakaid please review the updated changes. If we return an array of subsets wouldn't it make this function complex to use? The user of this function then has to deal with offsets if he wants to extract the strings. We can customize this function if we want to make it efficient by not duplicating the strings, but at present the usage from nanofi tailfile is to split blocks of chars of size 4KB. I don't think that is big enough to think of efficiency. As we move along and  encounter use cases we will refine this function and make it efficient.